### PR TITLE
CMake build system for VNote

### PR DESCRIPTION
**Pull request by miurahr**
CMake is a cross-platform free and open-source software (meta) build system for managing the build process of software using a compiler-independent method. 
Recent IDEs such as QtCreator, JetBrains CLion, VSCode, Visual Studio 2017, XCode supports CMake.

Here is a proposal to add CMake build script to VNote project.

Features
======

* Support building VNote project on Linux(GCC, CLANG), OS X(Apple Clang and Xcode) and Windows(MSVC).

* Support packaging.  DEB, RPM, TBZ2 and AppImage on Linux, DMG on Mac OS X, and Installer(NullSoft Installer System) and ZIP archive on WIndows.

* Support Qt Installer Framework (IFW)  to produce GUI installer for Linux, Mac OS X and Windows.

* Support statistic code analyzer;  clang-tidy, include-what-you-use and crazy check.

* Support and tested IDEs;  QtCreator, JetBrains CLion, Visual Studio 2017. 

Benefit
=======

* We can utilize modern IDEs to coding and refactoring.

* Improve packaging.

* It can co-exist with current QMake scripts.

Limitations and known issues
========================

* Qt 5.9 now support with MSVS2017(64bit), MSVS2015(32bit, 64bit), mingw-w64(32bit) on windows, and MSVC2017 only support CMake as IDE's native build system.  So we need to build 32bit binary using CMake with command line using VisualC++ bundled with MSVS2015.

* No test on Mac OS X (Because I don't have).

* Build exe successful but fails to build AppImage packaging in QtCreator on Linux

HOWTO BUILD
===========

* On Ubuntu Linux Bionic;

```bash
$ sudo apt-get install -y cmake g++ qt5-default
$ mkdir cmake-build-debug
$ cd cmake-build-debug
$ cmake  -DQt5_DIR=/usr/lib/x86_64-linux-gnu/cmake/Qt5/ /where/to/vnote/source/dir/
$ make -j3
$ make bundle
```

On Windows

* Install Qt5 and Visual Studio 2017 with CMake support
* Edit `CMakeSettings.json` for Qt5 path.
* Open CMakeLists.txt from VS2017.
* Run build all
* Run build `src/bundle` for packaging.

On Windows command line

```CMD
C:> mkdir build
C:> cd build
C:\build> cmake -G "Microsoft Visual Studio 2017 15 Win64"  -DQt5_DIR=C:\Qt\Qt5.9.7\5.9.7\msvc2017_64\lib\cmake\Qt5\ C:\where\vnote\source\
C:\build> cmake --build . --target all  --config Debug
C:\build> cmake --build . --target bundle
```

When running clang-tidy code checker, please configure `cmake -DCMake_RUN_CLANG_TIDY=ON ..`
also  `cmake -DCMake_RUN_CLAZY =ON ..` enables crazy checker and `-DCMake_RUN_IWYU=ON` enables `include-what-you-use `.

### CODE REVIEW

File: src/CMakeLists.txt [hunk 1]
```diff
@@ -0,0 +1,209 @@
+add_executable(VNote main.cpp
+    vmainwindow.cpp
```
Comment (tamlok) [inline src/CMakeLists.txt]: Indent this one properly?

Comment (miurahr) [reply src/CMakeLists.txt]: This indentation is wrong, it should be fixed.

**Comment by miurahr** [comment]
Note: It is necessary to install `windeployqt` as a preparation on Linux to procude AppImage package.

File: src/Packaging.cmake [hunk 2]
```diff
@@ -0,0 +1,179 @@
+
+find_package(Qt5Core REQUIRED)
+get_target_property(_qmake_executable Qt5::qmake IMPORTED_LOCATION)
+get_filename_component(_qt_bin_dir "${_qmake_executable}" DIRECTORY)
+find_program(WINDEPLOYQT_EXECUTABLE windeployqt HINTS "${_qt_bin_dir}")
+find_program(LINUXDEPLOYQT_EXECUTABLE linuxdeployqt HINTS "${_qt_bin_dir}")
```
Comment (miurahr) [inline src/Packaging.cmake]: Here I assumes `linuxdeployqt` is accecible as a name `linuxdeployqt` here. There may be a `linuxdeployqt-continuous-x86_64.AppImage`

**Comment by miurahr** [comment]
Now it is ready for merge.

**Pull request merged**